src/engine/Result.cpp

```diff
@@ -100,22 +171,130 @@ auto Result::getOrComputeDatatypeCountsPerColumn()
 }
 
 // _____________________________________________________________
-bool Result::checkDefinedness(const VariableToColumnMap& varColMap) {
-  const auto& datatypesPerColumn = getOrComputeDatatypeCountsPerColumn();
-  return std::ranges::all_of(varColMap, [&](const auto& varAndCol) {
-    const auto& [columnIndex, mightContainUndef] = varAndCol.second;
-    bool hasUndefined = datatypesPerColumn.at(columnIndex)
-                            .at(static_cast<size_t>(Datatype::Undefined)) != 0;
-    return mightContainUndef == ColumnIndexAndTypeInfo::PossiblyUndefined ||
-           !hasUndefined;
-  });
+void Result::checkDefinedness(const VariableToColumnMap& varColMap) {
+  auto performCheck = [](const auto& map, IdTable& idTable) {
+    DatatypeCountsPerColumn datatypeCountsPerColumn =
+        computeDatatypeCountsPerColumn(idTable);
+    return std::ranges::all_of(map, [&](const auto& varAndCol) {
+      const auto& [columnIndex, mightContainUndef] = varAndCol.second;
+      bool hasUndefined =
+          datatypeCountsPerColumn.at(columnIndex)
+              .at(static_cast<size_t>(Datatype::Undefined)) != 0;
+      return mightContainUndef == ColumnIndexAndTypeInfo::PossiblyUndefined ||
+             !hasUndefined;
+    });
+  };
+  if (isDataEvaluated()) {
+    AD_EXPENSIVE_CHECK(performCheck(varColMap, std::get<IdTable>(data_)));
+  } else {
+    auto generator = [](cppcoro::generator<IdTable> original,
+                        VariableToColumnMap varColMap,
+                        auto performCheck) -> cppcoro::generator<IdTable> {
+      bool first = true;
+      for (auto&& idTable : original) {
+        if (first) {
+          first = false;
+          // No need to check subsequent idTables assuming the datatypes
+          // don't change mid result.
+          AD_EXPENSIVE_CHECK(performCheck(varColMap, idTable));
+        }
+        co_yield std::move(idTable);
+      }
+    }(std::move(idTables()), varColMap, std::move(performCheck));
+    data_ = std::move(generator);
```

Comment:
I agree that the pattern is kind of very common here, but the individual functions are just individually too different that it would only make sense to unify 2 functions into a common interface here (`assertThatLimitWasRespected` and `checkDefinedness`), but the latter doesn't require any state inbetween the chunks so while it would be an acceptable abstraction layer for the former case, the latter would just consist out of 2 no-op lamdas which doesn't help for readability.